lib/integrations/awsoidc/eks_enroll_clusters.go

```diff
@@ -0,0 +1,472 @@
+	ident, err := identityGetter(ctx, nil)
+	if err != nil {
+		return "", trace.Wrap(err)
+
+	}
+
+	parsedARN, err := arn.Parse(aws.ToString(ident.Arn))
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	roleParts := strings.Split(parsedARN.Resource, "/")
+	if len(roleParts) < 2 {
+		return "", trace.BadParameter("can't get role name from arn %q", parsedARN.String())
+	}
+
+	return fmt.Sprintf("arn:aws:iam::%s:role/%s", parsedARN.AccountID, roleParts[1]), nil
+}
+
+// maybeAddAccessEntry checks list of access entries for the EKS cluster and adds one for Teleport if it's missing.
+// If access entry was added by this function it will return true as a first value.
+func maybeAddAccessEntry(ctx context.Context, clusterName, roleArn string, clt EnrollEKSCLusterClient) (bool, error) {
+	entries, err := clt.ListAccessEntries(ctx, &eks.ListAccessEntriesInput{
+		ClusterName: aws.String(clusterName),
+	})
+	if err != nil {
+		return false, trace.Wrap(err)
+	}
+
+	teleportEntryPresent := false
+	for _, entry := range entries.AccessEntries {
+		if entry == roleArn {
+			teleportEntryPresent = true
+			break
+		}
+	}
+
+	if !teleportEntryPresent {
+		_, err = clt.CreateAccessEntry(ctx, &eks.CreateAccessEntryInput{
+			ClusterName:  aws.String(clusterName),
+			PrincipalArn: aws.String(roleArn),
+		})
+		return err == nil, trace.Wrap(err)
+	}
+
+	return false, nil
+}
+
+// getPresignURL returns specially formatted URL that can be presigned and used in EKS authentication.
+func getPresignURL() url.URL {
+	endpoint := "sts.amazonaws.com"
+	q := url.Values{}
+	q.Set("Action", "GetCallerIdentity")
+	q.Set("Version", "2011-06-15")
+	q.Set("X-Amz-Expires", "60")
+
+	return url.URL{
+		Scheme:   "https",
+		Host:     endpoint,
+		Path:     "/",
+		RawQuery: q.Encode(),
+	}
+}
+
+// getKubeconfig returns kubeconfig that can be used to access target EKS cluster
+func getKubeconfig(ctx context.Context, timestamp time.Time, credsProvider aws.CredentialsProvider, clusterName, region, clusterCA, clusterEndpoint string) (*genericclioptions.ConfigFlags, error) {
+	targetUrl := getPresignURL()
+
+	r, err := http.NewRequest(http.MethodGet, targetUrl.String(), nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	r.Header.Add(awsHeaderClusterName, clusterName)
+	creds, err := credsProvider.Retrieve(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	signer := v4.NewSigner()
+	presigned, _, err := signer.PresignHTTP(ctx, creds, r, hashForGetRequests, "sts", region, timestamp)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	kubeToken := awsKubePrefix + base64.RawURLEncoding.EncodeToString([]byte(presigned))
+
+	eksClusterCA, err := base64.StdEncoding.DecodeString(clusterCA)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	restConfig := &rest.Config{
+		Host:        clusterEndpoint,
+		BearerToken: kubeToken,
+		TLSClientConfig: rest.TLSClientConfig{
+			CAData: eksClusterCA,
+		},
+	}
+
+	kubeConfig := genericclioptions.NewConfigFlags(false)
```

Comment:
rename it to something else than kubeConfig. When I read it the first time, I got the impression it was the kubeconfig but it's just an ConfigFlag structure